File: libs/shared-models/src/track.ts
import { Coordinate, TrackProcessingState } from './common';
import {
  GeodeticVelocity,
  GeodeticAcceleration,
  PositionCovariance,
  PositionVelocityCovariance,
  VelocityCovariance,
} from './kinematics';

export interface TrackPoint {
  id: string;
  entityId: string;
  position: Coordinate;
  heading?: number;
  speedKnots?: number;
  course?: number;
  source?: string;
  timestamp: string;
  altitude?: number;
  trackProcessingState?: TrackProcessingState;
  velocity?: GeodeticVelocity;
  acceleration?: GeodeticAcceleration;
  positionCovariance?: PositionCovariance;
  positionVelocityCovariance?: PositionVelocityCovariance;
  velocityCovariance?: VelocityCovariance;
  circularError?: number;
  altitudeError?: number;
  sensorId?: string;
}

export interface TrackSegment {
  segmentId: string;
  entityId: string;
  points: TrackPoint[];
  startTime: string;
  endTime: string;
}

export interface GetTrackHistoryRequest {
  entityId: string;
  startTime: string;
  endTime: string;
  maxPoints?: number;
  simplify?: boolean;
}

export interface GetTrackHistoryResponse {
  segments: TrackSegment[];
  totalPoints: number;
}

export interface TrackReplayRequest {
  entityId: string;
  startTime: string;
  endTime: string;
  speedMultiplier: number;
}

export interface TrackReplayEvent {
  entityId: string;
  point: TrackPoint;
  replayTimestamp: string;
}

File: libs/common-go/models/track.go
package models

// TrackPoint represents a single position observation for a tracked entity.
type TrackPoint struct {
	ID                         string                      `json:"id"`
	EntityID                   string                      `json:"entityId"`
	Position                   Coordinate                  `json:"position"`
	Heading                    *float64                    `json:"heading,omitempty"`
	SpeedKnots                 *float64                    `json:"speedKnots,omitempty"`
	Course                     *float64                    `json:"course,omitempty"`
	Source                     *string                     `json:"source,omitempty"`
	Timestamp                  string                      `json:"timestamp"`
	Altitude                   *float64                    `json:"altitude,omitempty"`
	TrackProcessingState       *string                     `json:"trackProcessingState,omitempty"`
	Velocity                   *GeodeticVelocity           `json:"velocity,omitempty"`
	Acceleration               *GeodeticAcceleration       `json:"acceleration,omitempty"`
	PositionCovariance         *PositionCovariance         `json:"positionCovariance,omitempty"`
	PositionVelocityCovariance *PositionVelocityCovariance `json:"positionVelocityCovariance,omitempty"`
	VelocityCovariance         *VelocityCovariance         `json:"velocityCovariance,omitempty"`
	CircularError              *float64                    `json:"circularError,omitempty"`
	AltitudeError              *float64                    `json:"altitudeError,omitempty"`
	SensorID                   *string                     `json:"sensorId,omitempty"`
}

// TrackSegment represents a contiguous segment of track points.
type TrackSegment struct {
	SegmentID string       `json:"segmentId"`
	EntityID  string       `json:"entityId"`
	Points    []TrackPoint `json:"points"`
	StartTime string       `json:"startTime"`
	EndTime   string       `json:"endTime"`
}
